### DISCUSSION TITLE
Unable to cast object exception

### DISCUSSION THREAD

**Discussion by FraserElectronics** [AvaloniaUI/Avalonia#13383] [Q&A] [answered]
I am struggling with a strange problem. I have created what amounts to a single page application with Avalonia. 
I have an About box that shares the same screen space as my main view controls but use the Z-Index to keep it behind the main view. I have the Z-Index of both the main view and About box view bound to properties on the main window view model.

The main view pseudo code:

```
	<Grid RowDefinitions="Auto">
		<Grid Grid.Row="0" ZIndex="{Binding MainViewZIndex}">
			<!-- All my main view controls here -->
		</Grid>
		<views:AboutView Grid.Row="0" ZIndex="{Binding AboutViewZIndex}"/>
	</Grid>
```

The AboutView is simply a user control.
During initial testing it worked great until I then bound a ViewModel to my AboutView user control.

So in my AboutView code-behind I have:

```
public partial class AboutView : UserControl
{
    public AboutView()
    {
        this.DataContext = new AboutViewViewModel();
        InitializeComponent();
    }
}
```

As soon as I run this and before anything is rendered, I get the exception:

```
Unable to cast object of type 'MyApp.Views.AboutView' to type 'MyApp.ViewModels.IMainWindowViewModel'.
```

If I remove the ```this.DataContext = ....``` from the AboutView, the application runs fine !

Anybody give me a clue as to what I am doing wrong?

Andy

**Comment by workgroupengineering** [comment]
If you can make a simple project that simulates the issue it is easier for someone to respond to you quickly.

**Comment by thevortexcloud** [comment]
Are you using a view locator or manually creating your views/view models? Because it almost sounds like you are doing something backwards.

**Reply by FraserElectronics** [reply]
Not view locators but I am using a service container for my view models (and other services).
However, the demo that I am about to upload to Github does not use those and still shows a problem.

Andy

**Comment by FraserElectronics** [comment]
Please see this repo that shows the issue:

https://github.com/FraserElectronics/AvaloniaCastException

If you comment out ```this.DataContext = new ViewOneViewModel();``` in ViewOne.axaml.cs and the same in ViewTwo.axaml.cs, the exceptions disappear and the code works.

I am sure I am doing something really dumb but can't see what it is :-(

Andy

**Reply by workgroupengineering** [reply]
To fix your problem, edit the xaml as follows:

```xaml
    <views:ViewOne Grid.Row="1" ZIndex="{Binding $parent.((vm:MainViewModel)DataContext).ViewOneZIndex}"/>
    <views:ViewTwo Grid.Row="1" ZIndex="{Binding $parent.((vm:MainViewModel)DataContext).ViewTwoZIndex}"/>
```

**Reply by FraserElectronics** [reply]
Are you able to explain what I was doing wrong?
It seems all my struggles with Avalonia come down to binding :-(

Andy

**Reply by timunie** [reply]
check docs about compiled bindings. https://docs.avaloniaui.net/docs/next/basics/data/data-binding/compiled-bindings 

Since 11.0 these are on by default iirc

**Reply by workgroupengineering** [reply, marked as answer]
When you set `this.DataContext = new ViewOneViewModel();` the DataContext of `{Binding ViewOneZIndex}` will be `ViewOneViewModel` instead of `vm:MainViewMode`. `{Binding $parent.((vm:MainViewModel)DataContext)` tells at Binding Engine to get the DataContext of first logical parent of `ViewOne` and then cast it to `vm:MainViewModel`.

**Reply by FraserElectronics** [reply]
Okay, I see where I am going wrong :-)
Thank you both for your help 👍

Andy
